### PR TITLE
Add unit tests to increase code coverage

### PR DESCRIPTION
**Pull request by soda480**
Added the following unit tests to get coverage up to 25%. 

My approach was to add unit tests without modifying the source code, however IMHO in order to unit test the remaining functions and methods will require refactoring the source code to make it **testable**. What I mean by testable is essentially refactoring the code to facilitate the mocking of  dependencies.

The patterns to implement to make the code more testable are described in the following article:
https://husobee.github.io/golang/testing/unit-test/2015/06/08/golang-unit-testing.html

- **Monkey Patching**
This pattern can be used for testing/mocking third party function calls.. Go has first class functions, thus assign the third party function to a variable defined at the package then in test override the variable to point to a function in your test package that returns the desired output.

- **Mocking Struct Methods via Interfacing**
This pattern can be used for testing/mocking methods of struct dependencies. Define an interface consisting of all methods that are called on the dependent struct. Include a get function that returns the interface (in lieu of returning the dependent struct). In test, create a struct that mocks the dependent struct and implement all the methods defined by the interface returning the desired (i.e. mocked) output. When overriding the get function in test return the address of the mock struct instead. Go will not complain because the mocked struct implements the same interface as in the actual code.

- **Using composition**
This pattern may be used to help us test the Extent struct. This involves composing larger interfaces from smaller ones through embedding, enabling us to control the methods which we do want to change while still being able to test the ones we don’t want to change.

I plan to review my findings as well as get recommendations from a resident Go developer, will use that output to create a new user story to refactor git-semver for testability user story.

The following repository contains sample code and unit tests showing how the patterns described above can be applied, writing sample code and unit tests made the concepts resonate with me: 
https://github.com/soda480/go-unit-test-example

The test fixture pattern used to mock the Worktree Filesystem was inspired by reviewing test code from: https://github.com/src-d/go-git/blob/v4.13.1/worktree_test.go#L157
I would have preferred to mock all filesystem access completely, but given how things were currently defined I wasn't sure how to do that without refactoring to use interfaces. However, being that others are using memory to mock filesystem this seemed like a reasonable alternative.

Signed-off-by: Emilio Reyes <emilio.reyes@intel.com>

### CODE REVIEW

**Comment by lranjbar** [comment]
I have an idea of how to run these tests... but can we add instructions on how to run the tests to the README.md?

File: scope/make_test.go [hunk 1]
```diff
@@ -165,3 +173,146 @@ func TestBumpPre(t *testing.T) {
...
+}
+
+func TestReadVersion_OpenError(t *testing.T) {
+	// ReadVersion Should return expected When Open returns an error
+	myExtentMock := &Extent{
+		Branch: "testbranch",
+	}
+	svExtentMock := &Extent{
+		Tree: &git.Worktree{
+			Filesystem: memfs.New(),
```
Comment (adam-intel) [inline scope/make_test.go]: I love the idea of using memfs to test this!

**Comment by adam-intel** [comment]
The use of memfs is an insightful way to test functions that use Extent.Tree.Filesystem.  This can be expanded to test the **Bump** function in a similar way that you tested **ReadVersion** just add the axis and pre.     

```
myExtentMock := &Extent{
	Branch: "testbranch",
}
	
memfs := memfs.New()
f, _ := memfs.Create("testbranch")
f.Write([]byte("v0.1.0"))

svExtentMock := &Extent{
	Tree: &git.Worktree{
		Filesystem: memfs,
	},
}
```

Also testing **WithPre** and **WithBuild** can be created because they don't require any external dependencies.  This should increase coverage by ~5% to ~10%.  

But, I agree with you getting tests to above 50% code coverage will likely require source code changes and abstracting the dependency of `gopkg.in/src-d/go-git.v4` into interface methods

**Review by adam-intel — COMMENTED** [review]
I am going to push back on this a little, I don't feel comfortable with the use of monkey patching in this code, I don't think it is adding any value.  I don't consider additional unit test % points worth it if it reducesses the readability of the code.   If the code needs to be restructured to be more testable then it should be restructured.  Here is a link about how to write golang code that uses external deps that maintains testability.  https://medium.com/swlh/using-go-interfaces-for-testable-code-d2e11b02dea 

Furthermore the ReadVersion function isn't an external dependency it's testable code that shouldn't be black boxed.

Thoughts? @jim-wang-intel

**Comment by jim-wang-yutsung** [comment]
I agree with @adam-intel.  We should refactor the original code if possible. 

I am in favor of using mock via interfaces.  That way the code is much more testable and maintainable.  Please try to avoid global variables if you can as they are bad since all functions has access to them and can become maintenance struggling, hard to figure out which functions actually use these variables.

**Comment by soda480** [comment]
thanks @adam-intel and @jim-wang-intel looks like I am back to the drawing board on this one, and I agree that the source code should be refactored to facilitate unit testing.
Do you guys have any recommendations for how to properly mock **internal** functions that don't involve monkey patching? for example how do I mock the MakeVersion call when unit testing the ReadVersion function?

**Comment by jim-wang-yutsung** [comment]
@soda480 I have not read all details yet but from what I have so far, we probably can start with refactoring the code with an interface, maybe called `Versioner`:
```go
type Versioner interface {
  Make(....)
  Read(...)
  Write(...)
}
```
and in the test code we can implement or override the signature of that `Versioner` interface to test it.

**Comment by jim-wang-yutsung** [comment]
Also, we use gomock to do the auto-generation of mocking code if necessary, for usage and references, you can see details in https://github.com/golang/mock#building-mocks.  
Alternatively, look at this to generate mocking code: https://github.com/vektra/mockery

**Pull request closed without merging**